diff --git a/src/dynarmic/backend/x64/a32_emit_x64.cpp b/src/dynarmic/backend/x64/a32_emit_x64.cpp
--- a/src/dynarmic/backend/x64/a32_emit_x64.cpp
+++ b/src/dynarmic/backend/x64/a32_emit_x64.cpp
@@ -232 +232 @@
-    Xbyak::Label pass = EmitCond(ctx.block.GetCondition());
+    Xbyak::Label pass = EmitCond(ctx.block.GetCondition(), code.T_NEAR);
@@ -1362 +1362 @@
-    Xbyak::Label pass = EmitCond(terminal.if_);
+    Xbyak::Label pass = EmitCond(terminal.if_, code.T_NEAR);
@@ -1383,11 +1383,19 @@ void A32EmitX64::EmitPatchJg(const IR::LocationDescriptor& target_desc, CodePtr target_code_ptr) {
     const CodePtr patch_location = code.getCurr();
     if (conf.fast_dispatch_table_link && conf.HasOptimization(OptimizationFlag::FastDispatch)) {
-        code.jg(terminal_handler_fast_dispatch_hint);
+        Xbyak::Label stable_link, after_stable_link;
+        code.jg(stable_link, code.T_NEAR);
+        code.EnsurePatchLocationSize(patch_location, 14);
+        code.jmp(after_stable_link, code.T_NEAR);
+        code.L(stable_link);
+        EmitStableLink(target_desc);
+        code.L(after_stable_link);
     } else if (target_code_ptr) {
         code.jg(target_code_ptr);
     } else {
         code.mov(MJitStateReg(A32::Reg::PC), A32::LocationDescriptor{target_desc}.PC());
         code.jg(code.GetReturnFromRunCodeAddress());
     }
-    code.EnsurePatchLocationSize(patch_location, 14);
+    if (!(conf.fast_dispatch_table_link && conf.HasOptimization(OptimizationFlag::FastDispatch))) {
+        code.EnsurePatchLocationSize(patch_location, 14);
+    }
 }
@@ -1396,11 +1403,19 @@ void A32EmitX64::EmitPatchJz(const IR::LocationDescriptor& target_desc, CodePtr target_code_ptr) {
     const CodePtr patch_location = code.getCurr();
     if (conf.fast_dispatch_table_link && conf.HasOptimization(OptimizationFlag::FastDispatch)) {
-        code.jz(terminal_handler_fast_dispatch_hint);
+        Xbyak::Label stable_link, after_stable_link;
+        code.jz(stable_link, code.T_NEAR);
+        code.EnsurePatchLocationSize(patch_location, 14);
+        code.jmp(after_stable_link, code.T_NEAR);
+        code.L(stable_link);
+        EmitStableLink(target_desc);
+        code.L(after_stable_link);
     } else if (target_code_ptr) {
         code.jz(target_code_ptr);
     } else {
         code.mov(MJitStateReg(A32::Reg::PC), A32::LocationDescriptor{target_desc}.PC());
         code.jz(code.GetReturnFromRunCodeAddress());
     }
-    code.EnsurePatchLocationSize(patch_location, 14);
+    if (!(conf.fast_dispatch_table_link && conf.HasOptimization(OptimizationFlag::FastDispatch))) {
+        code.EnsurePatchLocationSize(patch_location, 14);
+    }
 }
diff --git a/src/dynarmic/backend/x64/emit_x64.h b/src/dynarmic/backend/x64/emit_x64.h
--- a/src/dynarmic/backend/x64/emit_x64.h
+++ b/src/dynarmic/backend/x64/emit_x64.h
@@ -107 +107,3 @@
-    Xbyak::Label EmitCond(IR::Cond cond);
+    Xbyak::Label EmitCond(
+        IR::Cond cond,
+        Xbyak::CodeGenerator::LabelType label_type = Xbyak::CodeGenerator::T_AUTO);
diff --git a/src/dynarmic/backend/x64/emit_x64.cpp b/src/dynarmic/backend/x64/emit_x64.cpp
--- a/src/dynarmic/backend/x64/emit_x64.cpp
+++ b/src/dynarmic/backend/x64/emit_x64.cpp
@@ -279 +279,2 @@
-Xbyak::Label EmitX64::EmitCond(IR::Cond cond) {
+Xbyak::Label EmitX64::EmitCond(
+        IR::Cond cond, Xbyak::CodeGenerator::LabelType label_type) {
@@ -286,45 +287,45 @@
     switch (cond) {
     case IR::Cond::EQ:
-        code.jz(pass);
+        code.jz(pass, label_type);
         break;
     case IR::Cond::NE:
-        code.jnz(pass);
+        code.jnz(pass, label_type);
         break;
     case IR::Cond::CS:
-        code.jc(pass);
+        code.jc(pass, label_type);
         break;
     case IR::Cond::CC:
-        code.jnc(pass);
+        code.jnc(pass, label_type);
         break;
     case IR::Cond::MI:
-        code.js(pass);
+        code.js(pass, label_type);
         break;
     case IR::Cond::PL:
-        code.jns(pass);
+        code.jns(pass, label_type);
         break;
     case IR::Cond::VS:
-        code.jo(pass);
+        code.jo(pass, label_type);
         break;
     case IR::Cond::VC:
-        code.jno(pass);
+        code.jno(pass, label_type);
         break;
     case IR::Cond::HI:
-        code.ja(pass);
+        code.ja(pass, label_type);
         break;
     case IR::Cond::LS:
-        code.jna(pass);
+        code.jna(pass, label_type);
         break;
     case IR::Cond::GE:
-        code.jge(pass);
+        code.jge(pass, label_type);
         break;
     case IR::Cond::LT:
-        code.jl(pass);
+        code.jl(pass, label_type);
         break;
     case IR::Cond::GT:
-        code.jg(pass);
+        code.jg(pass, label_type);
         break;
     case IR::Cond::LE:
-        code.jle(pass);
+        code.jle(pass, label_type);
         break;
     default:
         ASSERT_MSG(false, "Unknown cond {}", static_cast<size_t>(cond));
